src/Symfony/Component/Clock/MockClock.php

```diff
@@ -38,23 +43,29 @@ public function __construct(\DateTimeImmutable|string $now = 'now', \DateTimeZon
             }
         }
 
+        $this->timezone = $timezone;
+
         if (\is_string($now)) {
-            $now = new DatePoint($now, $timezone ?? new \DateTimeZone('UTC'));
-        } elseif (!$now instanceof DatePoint) {
+            $now = new DatePoint($now, $this->timezone ?? new \DateTimeZone('UTC'));
+        } elseif (!$now instanceof DatePoint && !\is_callable($now)) {
             $now = DatePoint::createFromInterface($now);
         }
 
-        $this->now = null !== $timezone ? $now->setTimezone($timezone) : $now;
+        if (\is_callable($now)) {
+            $this->timeProvider = $now;
+        } else {
+            $this->now = null !== $this->timezone ? $now->setTimezone($this->timezone) : $now;
+        }
     }
 
     public function now(): DatePoint
     {
-        return clone $this->now;
+        return clone ($this->now ??= $this->tryFromTimeProvider());
```

Comment:
what about removing the method? it's mostly boilerplate, no?
```suggestion
        return clone $this->now ??= ($this->timeProvider)();
```

Reply:
For the record: we keep the method because it also take care of the timezone.